### PR TITLE
[Pefize users]

### PR DESCRIPTION
**Pull request by pierallard**
Remaining:

- [x] Avatar
- [x] Migration for avatars
- [x] Remove show
- [x] Update tests to be sure the everybody can not update every users
- [x] Remove dead code
- [x] Update current user does not work
- [x] EE features to be put in EE (check diffs demo/this PR)
- [x] Translate everything
- [x] Product Grid Filters 
- [x] ProductGrid filters should include system ones vendor/akeneo/pim-community-dev/src/Pim/Bundle/EnrichBundle/Form/Type/ProductGridFilterChoiceType.php
- [x] ProductGrid filters should only return usable as grid filters
- [x] ProductGrid filters are different in EE to add "Permissions"
- [x] Move edit/show fields from form_extensions into right BC
- [x] Ensure edit/show are synchronized (between files)
- [x] REMOVE THE USERPREFERENCE SUBSCRIBER (Ctrl+f TODO Remove)
- [x] Execute JS things on upgrade.md

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Todo
| Added legacy Behats               | Todo
| Added acceptance tests            | Todo
| Added integration tests           | Todo
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Akeneo/UserManagement/Component/Normalizer/UserNormalizer.php [hunk 1]
```diff
@@ -14,34 +15,50 @@
  */
 class UserNormalizer implements NormalizerInterface
 {
+    /** @var DateTimeNormalizer */
+    private $dateTimeNormalizer;
+
     /** @var array */
     protected $supportedFormats = ['array', 'standard', 'internal_api'];
 
+    public function __construct(DateTimeNormalizer $dateTimeNormalizer)
```
Comment (pierallard) [inline src/Akeneo/UserManagement/Component/Normalizer/UserNormalizer.php]: construct

File: src/Pim/Bundle/EnrichBundle/Resources/public/templates/form/common/fields/media.html [hunk 2]
```diff
@@ -0,0 +1,37 @@
+<div class="AknMediaField<%- media.filePath ? ' has-file' : '' %><%- readOnly ? ' AknMediaField--disabled' : '' %>" >
+    <% if (media.filePath === null) { %>
+        <input class="AknMediaField-fileUploaderInput<%- readOnly ? ' AknMediaField-fileUploaderInput--disabled' : '' %>" type="file" />
+        <div class="AknMediaField-emptyContainer">
+            <span title="upload icon" class="AknMediaField-uploadIcon"/>
+            <span>Upload</span>
```
Comment (pierallard) [inline src/Pim/Bundle/EnrichBundle/Resources/public/templates/form/common/fields/media.html]: upload

File: src/Akeneo/UserManagement/Bundle/Resources/config/requirejs.yml [hunk 3]
```diff
@@ -7,9 +7,45 @@ config:
...
                     options:
                         urls:
                             list: pim_user_user_group_rest_index
-                            search: pim_user_user_group_rest_index
+                user-role:
+                    module: pim/base-fetcher
+                    options:
+                        urls:
+                            list: pim_user_user_role_rest_index
+        pim/controller-registry:
```
Comment (pierallard) [inline src/Akeneo/UserManagement/Bundle/Resources/config/requirejs.yml]: this thing has probably moved

File: src/Akeneo/Pim/Enrichment/Bundle/Controller/InternalApi/ProductGridFilterController.php [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+     * Return the filter configured in the grid ($datagridName)
+     *
+     * @param string      $locale
+     * @param string|null $search
+     * @param int         $limit
+     * @param int         $page
+     *
+     * @return array
+     */
+    private function getSystemFilters(
```
Comment (pierallard) [inline src/Akeneo/Pim/Enrichment/Bundle/Controller/InternalApi/ProductGridFilterController.php]: @anaelChardan

File: src/Akeneo/UserManagement/Bundle/Resources/config/requirejs.yml (hunk 3 above)
Comment (anaelChardan) [reply src/Akeneo/UserManagement/Bundle/Resources/config/requirejs.yml]: look to remove one

File: src/Pim/Bundle/EnrichBundle/Resources/public/templates/form/common/fields/media.html (hunk 2 above)
Comment (anaelChardan) [reply src/Pim/Bundle/EnrichBundle/Resources/public/templates/form/common/fields/media.html]: translate it

File: src/Akeneo/UserManagement/Component/Normalizer/UserNormalizer.php (hunk 1 above)
Comment (anaelChardan) [reply src/Akeneo/UserManagement/Component/Normalizer/UserNormalizer.php]: comment on construct to have PHPDoc

File: src/Akeneo/Pim/Enrichment/Bundle/Controller/InternalApi/ProductGridFilterController.php (hunk 4 above)
Comment (anaelChardan) [reply src/Akeneo/Pim/Enrichment/Bundle/Controller/InternalApi/ProductGridFilterController.php]: To add Permission Filter on EE

File: src/Akeneo/UserManagement/Bundle/Resources/public/js/fields/default-product-grid-view.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+/**
+ * @author    Pierre Allard <pierre.allard@akeneo.com>
+ * @copyright 2018 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
+ */
+class InterfaceNormalizedDatagridView {
+  id: string;
+  label: string;
+  owner_id: number;
+  type: string; // 'public'|'project'
```
Comment (pierallard) [inline src/Akeneo/UserManagement/Bundle/Resources/public/js/fields/default-product-grid-view.ts]: real type

File: src/Akeneo/UserManagement/Bundle/Resources/public/js/fields/default-product-grid-view.ts [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+const __ = require('oro/translator');
+const BaseSelect = require('pim/form/common/fields/select');
+const FetcherRegistry = require('pim/fetcher-registry');
+
+/**
+ * @author    Pierre Allard <pierre.allard@akeneo.com>
+ * @copyright 2018 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
+ */
+class InterfaceNormalizedDatagridView {
```
Comment (pierallard) [inline src/Akeneo/UserManagement/Bundle/Resources/public/js/fields/default-product-grid-view.ts]: interface

File: src/Akeneo/UserManagement/Bundle/Resources/public/js/fields/product-grid-filters.ts [hunk 7]
```diff
@@ -0,0 +1,126 @@
+import * as $ from 'jquery';
+import * as i18n from 'pimui/js/i18n';
+import * as _ from "underscore";
```
Comment (pierallard) [inline src/Akeneo/UserManagement/Bundle/Resources/public/js/fields/product-grid-filters.ts]: simple quote

File: src/Akeneo/UserManagement/Bundle/Resources/public/js/fields/product-grid-filters.ts [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+        )
+      }
+    };
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  protected select2InitSelection(element: any, callback: any): void {
+    const strValues:string = (<any> $(element)).val();
```
Comment (pierallard) [inline src/Akeneo/UserManagement/Bundle/Resources/public/js/fields/product-grid-filters.ts]: as string

File: src/Akeneo/UserManagement/Bundle/Resources/public/js/fields/scope.ts [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+ * @author    Pierre Allard <pierre.allard@akeneo.com>
+ * @copyright 2018 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
+ */
+class InterfaceNormalizedChannel {
+  code: string;
+  labels: { [key:string] : string };
+}
+
+class ScopeField extends BaseSelect {
```
Comment (pierallard) [inline src/Akeneo/UserManagement/Bundle/Resources/public/js/fields/scope.ts]: ChannelField

File: src/Akeneo/UserManagement/Bundle/Resources/public/js/fields/scope.ts [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+const BaseSelect = require('pim/form/common/fields/select');
+const FetcherRegistry = require('pim/fetcher-registry');
+const UserContext = require('pim/user-context');
+
+/**
+ * @author    Pierre Allard <pierre.allard@akeneo.com>
+ * @copyright 2018 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
+ */
+class InterfaceNormalizedChannel {
```
Comment (pierallard) [inline src/Akeneo/UserManagement/Bundle/Resources/public/js/fields/scope.ts]: clss Interface

File: src/Akeneo/UserManagement/Bundle/Resources/public/js/fields/user-role.ts [hunk 11]
```diff
@@ -0,0 +1,41 @@
+import * as $ from 'jquery';
+const BaseSelect = require('pim/form/common/fields/select');
+const FetcherRegistry = require('pim/fetcher-registry');
+
+/**
+ * @author    Pierre Allard <pierre.allard@akeneo.com>
+ * @copyright 2018 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
+ */
+class InterfaceNormalizedUserRole {
```
Comment (pierallard) [inline src/Akeneo/UserManagement/Bundle/Resources/public/js/fields/user-role.ts]: /class Interface*/

File: tests/legacy/features/Context/DataGridContext.php [hunk 12]
```diff
@@ -504,12 +504,9 @@ public function theRowsShouldBeSortedBy($naturally, $order, $columnName)
      */
     public function iClickOnTheActionOfTheRowWhichContains($actionName, $element)
     {
-        // Temporary solution waiting for Category tree moving (PIM-6574)
-        $this->getSession()->executeScript('$(".AknDefault-mainContent").scrollLeft(1000)');
-
+        $this->getSession()->wait(6000, false); //TODO REMOVE THIS UGLY THING
```
Comment (pierallard) [inline tests/legacy/features/Context/DataGridContext.php]: TODO

File: tests/legacy/features/pim/enrichment/product/datagrid/datagrid_views.feature [hunk 13]
```diff
@@ -118,30 +118,28 @@ Feature: Datagrid views
...
+    When I edit the "Mary" user
+    Then I should see the text "Mary"
     And I should see the text "Save"
     When I visit the "Additional" tab
     Then I should see the text "Default product grid view"
     And I fill in the following information:
       | Default product grid view | Sneakers only |
     And I press the "Save" button
     Then I should not see the text "There are unsaved changes."
+    And I wait 5 seconds
```
Comment (pierallard) [inline tests/legacy/features/pim/enrichment/product/datagrid/datagrid_views.feature]: Nanou please fix this

File: tests/legacy/features/pim/enrichment/product/datagrid/datagrid_views.feature [hunk 14]
```diff
@@ -118,30 +118,28 @@ Feature: Datagrid views
...
-    Then I should see the text "Julia Stark"
-    And I should see the text "Save"
+    When I edit the "Julia" user
     When I visit the "Additional" tab
     Then I should see the text "Default product grid view"
     And I fill in the following information:
       | Default product grid view | Sneakers only |
     And I press the "Save" button
     Then I should not see the text "There are unsaved changes."
+    And I wait 5 seconds
```
Comment (pierallard) [inline tests/legacy/features/pim/enrichment/product/datagrid/datagrid_views.feature]: same

File: src/Akeneo/Platform/Bundle/UIBundle/Resources/public/templates/form/common/fields/select.html [hunk 15]
```diff
@@ -7,12 +7,27 @@
...
-    <% _.each(choices, function (choiceLabel, choiceValue) { %>
-        <option
-            value="<%- choiceValue %>"
-            <% if ((!multiple && choiceValue === value) || (multiple && _.contains(value, choiceValue))) { %> selected<% } %>
-        >
-            <%- choiceLabel %>
-        </option>
-    <% }); %>
+    <% if (Object.keys(choices).length > 0 && Object.values(choices)[0] instanceof Object) { %>
+        <% _.each(choices, function (optGroupOptions, optGroupLabel) { %>
```
[the remark this answers is not in the document]
Comment (anaelChardan) [reply src/Akeneo/Platform/Bundle/UIBundle/Resources/public/templates/form/common/fields/select.html]: SHOULD BE OK

[the remark this answers is not in the document]
Comment (anaelChardan) [reply src/Akeneo/Platform/Bundle/UIBundle/Resources/public/templates/form/common/fields/select.html]: does not work as they are objects

File: src/Akeneo/Platform/Bundle/UIBundle/Resources/public/js/form/common/fields/media.ts [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+        if (myXhr.upload) {
+          myXhr.upload.addEventListener('progress', this.handleProcess.bind(this), false);
+        }
+
+        return myXhr;
+      }
+    }).done((data) => {
+      this.setUploadContextValue(data);
+      this.render();
+    }).fail((xhr) => {
```
[the remark this answers is not in the document]
Comment (anaelChardan) [reply src/Akeneo/Platform/Bundle/UIBundle/Resources/public/js/form/common/fields/media.ts]: Does not cry, see that on master

File: docker-compose.yml [hunk 17]
```diff
@@ -55,7 +55,7 @@ services:
       - 'behat'
 
   mysql:
-    image: 'mysql:5.7'
+    image: 'mysql:5.7.22'
```
Comment (damien-carcel) [inline docker-compose.yml]: Nope. This is not related to the PR. Plus, you can do a `docker-compose pull` to have the latest version.
We need 5.7.22 at least, not 5.7.22 only.

File: src/Akeneo/Platform/Bundle/UIBundle/Resources/public/js/attribute/form/properties/available-locales.js [hunk 18]
```diff
@@ -1,95 +1,24 @@
-/**
- * @author    Yohan Blain <yohan.blain@akeneo.com>
- * @copyright 2017 Akeneo SAS (http://www.akeneo.com)
+ /**
```
Comment (damien-carcel) [inline src/Akeneo/Platform/Bundle/UIBundle/Resources/public/js/attribute/form/properties/available-locales.js]: useless space

File: src/Akeneo/Platform/Bundle/UIBundle/Resources/public/js/form/common/attributes.js [hunk 19]
```diff
@@ -264,8 +264,8 @@ define(
                         AttributeManager.isOptional(field.attribute, object)
                     );
                 }).then(function (field, channels, isOptional) {
-                    const scope = _.findWhere(channels, { code: UserContext.get('catalogScope') });
-                    const locale = UserContext.get('catalogLocale');
+                    var scope = _.findWhere(channels, { code: UserContext.get('catalogScope') });
```
Comment (damien-carcel) [inline src/Akeneo/Platform/Bundle/UIBundle/Resources/public/js/form/common/attributes.js]: why ` var` and not `const`?

File: src/Akeneo/Platform/Bundle/UIBundle/Resources/public/js/form/common/fields/available-locales.ts [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+type InterfaceNormalizedLocale = {
+  code: string;
+  label: string;
+}
+
+class AvailableLocales extends BaseSelect {
+  /**
+   * {@inheritdoc}
+   */
+  configure() {
```
Comment (damien-carcel) [inline src/Akeneo/Platform/Bundle/UIBundle/Resources/public/js/form/common/fields/available-locales.ts]: method visibility

File: src/Akeneo/Platform/Bundle/UIBundle/Resources/public/js/form/common/fields/media.ts [hunk 21]
```diff
@@ -0,0 +1,122 @@
...
+
+/**
+ * Media field (using back-end FileInfo)
+ *
+ * @author    Marie Gautier <marie.gautier@akeneo.com>
+ * @copyright 2018 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
+ */
+class MediaField extends BaseField {
+  readonly template = _.template(template);
```
Comment (damien-carcel) [inline src/Akeneo/Platform/Bundle/UIBundle/Resources/public/js/form/common/fields/media.ts]: visibility?

File: src/Akeneo/UserManagement/Bundle/Resources/config/factories.yml [hunk 22]
```diff
@@ -1,8 +1,12 @@
...
     pim_user.factory.user:
-        class: '%akeneo_storage_utils.factory.simple.class%'
+        class: 'Akeneo\UserManagement\Component\Factory\UserFactory'
         arguments:
-            - '%pim_user.entity.user.class%'
+            - '@pim_catalog.repository.locale'
+            - '@pim_catalog.repository.channel'
+            - '@pim_catalog.repository.category'
+            - '@pim_user.repository.group'
+            - 'Akeneo\UserManagement\Component\Model\User'
```
Comment (damien-carcel) [inline src/Akeneo/UserManagement/Bundle/Resources/config/factories.yml]: Why not using the class parameter here?
By hardcoding the namespace, you force integrators to redefine this service when they maybe just want to add a new method to the User model.

File: src/Akeneo/Platform/Bundle/UIBundle/Resources/public/js/attribute/form/properties/available-locales.js (hunk 18 above)
Comment (jjanvier) [reply src/Akeneo/Platform/Bundle/UIBundle/Resources/public/js/attribute/form/properties/available-locales.js]: no comments like that
it passes the CI, we have the step `back-code-style` to enforce the style is the one chosen by the team
there is no need for more

File: src/Akeneo/UserManagement/Bundle/Resources/config/factories.yml (hunk 22 above)
Comment (anaelChardan) [reply src/Akeneo/UserManagement/Bundle/Resources/config/factories.yml]: Because we don't want to do extensibility by overriding class parameters we are trying to remove all of them

File: src/Akeneo/Platform/Bundle/UIBundle/Resources/public/js/form/common/fields/available-locales.ts (hunk 20 above)
Comment (anaelChardan) [reply src/Akeneo/Platform/Bundle/UIBundle/Resources/public/js/form/common/fields/available-locales.ts]: public as it is the default visibility in TypeScript

File: src/Akeneo/Platform/Bundle/UIBundle/Resources/public/js/form/common/fields/media.ts (hunk 21 above)
Comment (anaelChardan) [reply src/Akeneo/Platform/Bundle/UIBundle/Resources/public/js/form/common/fields/media.ts]: public as it is the default visibility in TypeScript

File: docker-compose.yml (hunk 17 above)
Comment (damien-carcel) [reply docker-compose.yml]: @anaelChardan could you please answer my comment rather than just hiding it?

Comment (anaelChardan) [reply docker-compose.yml]: @damien-carcel Yes ;) I've corrected it.

File: src/Akeneo/Platform/Bundle/UIBundle/Resources/public/js/form/common/attributes.js (hunk 19 above)
Comment (anaelChardan) [reply src/Akeneo/Platform/Bundle/UIBundle/Resources/public/js/form/common/attributes.js]: @damien-carcel Yes ;) I've corrected it.

File: docker-compose.yml (hunk 17 above)
Comment (damien-carcel) [reply docker-compose.yml]: :+1:
